---
title: "Final Project R Code"
author: "Marilena Papavassiliou"
date: "2024-04-02"
output: html_document
  toc: true
---

# STAT 555 FINAL PROJECT DESCRIPTION

This project involves using RNA-seq and ATAC-seq data from four mouse cell lineages in different stages of hematopoiesis. This part of the project involves analyzing the RNA-seq data using the DEseq2 software. 

The four mouse cell lineages are: 
1. HSC - hematopoietic stem cell
2. CMP - common myeloid progenitor
3. CFU-E - colony forming unit erythroid
4. ERY - erythroblast

The data consists of two biological replicates for each cell lineage. 

The data is to be analyzed in a pairwise fashion.We need to at least analyze pairs 1&3, 1&4, and 1&2. For additional points we can analyze pairs 2&4, 3&4, and 2&3. 

We aim to answer the following questions for the RNA-seq part of the project: 
1. What genes are differentially expressed across each pair of cell lines?
2. What are the functions of the genes with differential expression patterns? (need to do GO-term analysis)
3. How consistent are the results between DEseq2 and limma voom?
4. Construct a hierarchical tree using all the RNA-seq data, and perform clustering analysis. Describe the relationship based on your results.

# Pairwise Comparison 1 & 3 - Hematopoietic Stem Cell VS Colony Forming Unit Erythroid

The first step for DEseq2 is importing a count matrix, with one row for each gene and one column for each sample. The cells in the count matrix give the number of reads that have been unambiguously mapped to a particular gene (DEseq2 paper). 

The TSV files I am working with were created with the software RSEM (mentioned on "Project Description" page on Canvas). We don't have access to raw counts, but we can use the column 'expected counts' in these files. 

```{r generate-count-matrix}

#read in files associated with hsc and cfu-e
hsc.1 <- read.delim("ENCFF247FEJ.tsv", header = TRUE, sep = "\t")
hsc.2 <- read.delim("ENCFF064MKY.tsv", header = TRUE, sep = "\t")
cfu.1 <- read.delim("ENCFF667IDY.tsv", header = TRUE, sep = "\t")
cfu.2 <- read.delim("ENCFF655LMK.tsv", header = TRUE, sep = "\t")

#extract gene id names and TPM columns from each table
count.matrix <- cbind(hsc.1$expected_count, hsc.2$expected_count, cfu.1$expected_count, cfu.2$expected_count)
#add gene ids as row names to the matrix
rownames(count.matrix) <- hsc.1$gene_id

#make a data frame of column names
col.names <- data.frame(row.names = c("HSC1", "HSC2", "CFU-E1", "CFU-E2"), condition = c("HSC", "HSC", "CFU-E1", "CFU-E1"))

#read in the count matrix as a DESeqDataSet
library("DESeq2")
dds <- DESeqDataSetFromMatrix(countData = round(count.matrix),
                              colData = col.names,
                              design = ~ condition) #regress on condition, so we compare HSC to CFU-E

```

Now we run differential expression analysis:

```{r differential-expression-1v3}

#perform DE analysis 
dds <- DESeq(dds)

#can get a summary of results here
res <- results(dds)
summary(res)

```

Dispersion of the log-fold-change (LFC) value may be artifically high between replicates of the same gene due to low sample size. We correct for this by performing an LFC shrinkage. 

```{r log-fold-change-shrinkage}

#find the name of the results you want to perform an LFC shrink on
resultsNames(dds) #we want "condition_HSC_vs_CFU.E1"

#perform LFC shrinkage
library("apeglm")
resLFC <- lfcShrink(dds, coef="condition_HSC_vs_CFU.E1", type="apeglm") #documentation indicates apeglm outperforms the normal shrinkage estimator

#get summary of output
summary(resLFC)

```

DEseq2 automatically performs a multiple testing correction on our p-values. Let's find the number of genes for which the adjusted p-value falls below alpha = 0.05

```{r retrieve-DE-gene-count}

sum(resLFC$padj < 0.05, na.rm = TRUE) #6923 genes are differentially expressed 

```

We can also generate an MA plot, with genes with adjusted p-values below 0.05 highlighted in blue. Triangles indicate genes that fall outside the visualized window of the plot. 

```{r MA-plot}

plotMA(resLFC, alpha = 0.05, main = "MA plot: HSC VS CFU")

```

We can export our significant results to a CSV file:

```{r export-results}

#subset by only the significant p-values 
resSig <- subset(resLFC, padj < 0.05)

#write to CSV file
write.csv(as.data.frame(resSig), 
          file="DE_HSC_vs_CFU.csv")

```

# Clustering Analysis

When performing clustering analysis, it's important to work with transformed versions of the count data. Here we use a variance-stabilizing transformation (VST), which corrects for the trend that variance increases with lower mean. DEseq2 also has another transformation which performs a similar function (rlog), but can be too slow when working with large (> hundreds of samples) data sets. 

We set the blind parameter in the below transformation to false. This is parameter specifies whether the transformation should be 'blind' to the values passed to the design formula. This is appropriate when you want to compare samples without adding the 'bias' of experimental groups - but here, we are interested in the differences between experimental groups (cell lineages), and so we set blind to false.

```{r clustering-analysis}

#perform variance stabilizing transformation 
vsd <- vst(dds, blind=FALSE)

#get sample-to-sample distances in a distance matrix
sampleDists <- dist(t(assay(vsd)))

#construct a heatmap from this distance matrix
library("pheatmap")
library("RColorBrewer")
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(vsd$condition, vsd$type, sep="-")
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)

```

Now repeat these analyses for all the other pairs. 

# Pairwise Comparison 1 & 4 - Hematopoietic Stem Cell VS Erythroblast

```{r DE-1v4}

#read in libraries
library("DESeq2")
library("apeglm")
library("pheatmap")
library("RColorBrewer")

#read in files associated with all cell lines
hsc.1 <- read.delim("ENCFF247FEJ.tsv", header = TRUE, sep = "\t")
hsc.2 <- read.delim("ENCFF064MKY.tsv", header = TRUE, sep = "\t")
ery.1 <- read.delim("ENCFF342WUL.tsv", header = TRUE, sep = "\t")
ery.2 <- read.delim("ENCFF858JHF.tsv", header = TRUE, sep = "\t")

#extract gene id names and TPM columns from each table
count.matrix <- cbind(hsc.1$expected_count, hsc.2$expected_count, ery.1$expected_count, ery.2$expected_count)
#add gene ids as row names to the matrix
rownames(count.matrix) <- hsc.1$gene_id

#make a data frame of column names
col.names <- data.frame(row.names = c("HSC1", "HSC2", "ERY1", "ERY2"), condition = c("HSC", "HSC", "ERY", "ERY"))

#read in the count matrix as a DESeqDataSet
dds <- DESeqDataSetFromMatrix(countData = round(count.matrix),
                              colData = col.names,
                              design = ~ condition) 

#perform DE analysis 
dds <- DESeq(dds)

#find the name of the results you want to perform an LFC shrink on
resultsNames(dds) 

#perform LFC shrinkage
resLFC <- lfcShrink(dds, coef="condition_HSC_vs_ERY", type="apeglm") #documentation indicates apeglm outperforms the normal shrinkage estimator

#subset by only the significant p-values 
resSig <- subset(resLFC, padj < 0.05)

#write to CSV file
write.csv(as.data.frame(resSig), 
          file="DE_HSC_vs_ERY.csv")

#make an MA plot
plotMA(resLFC, alpha = 0.05, main = "MA plot: HSC VS ERY")

#do clustering analysis 

#perform variance stabilizing transformation 
vsd <- vst(dds, blind=FALSE)

#get sample-to-sample distances in a distance matrix
sampleDists <- dist(t(assay(vsd)))

#construct a heatmap from this distance matrix
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(vsd$condition, vsd$type, sep="-")
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)

```

# Pairwise Comparison 1 & 2 - Hematopoietic Stem Cell VS Common Myeloid Progenitor

```{r DE-1v2}

#read in libraries
library("DESeq2")
library("apeglm")
library("pheatmap")
library("RColorBrewer")

#read in files associated with all cell lines
hsc.1 <- read.delim("ENCFF247FEJ.tsv", header = TRUE, sep = "\t")
hsc.2 <- read.delim("ENCFF064MKY.tsv", header = TRUE, sep = "\t")
cmp.1 <- read.delim("ENCFF623OLU.tsv", header = TRUE, sep = "\t")
cmp.2 <- read.delim("ENCFF691MHW.tsv", header = TRUE, sep = "\t")

#extract gene id names and TPM columns from each table
count.matrix <- cbind(hsc.1$expected_count, hsc.2$expected_count, cmp.1$expected_count, cmp.2$expected_count)
#add gene ids as row names to the matrix
rownames(count.matrix) <- hsc.1$gene_id

#make a data frame of column names
col.names <- data.frame(row.names = c("HSC1", "HSC2", "CMP1", "CMP2"), condition = c("HSC", "HSC", "CMP", "CMP"))

#read in the count matrix as a DESeqDataSet
dds <- DESeqDataSetFromMatrix(countData = round(count.matrix),
                              colData = col.names,
                              design = ~ condition) 

#perform DE analysis 
dds <- DESeq(dds)

#find the name of the results you want to perform an LFC shrink on
resultsNames(dds) 

#perform LFC shrinkage
resLFC <- lfcShrink(dds, coef="condition_HSC_vs_CMP", type="apeglm") #documentation indicates apeglm outperforms the normal shrinkage estimator

#subset by only the significant p-values 
resSig <- subset(resLFC, padj < 0.05)

#write to CSV file
write.csv(as.data.frame(resSig), 
          file="DE_HSC_vs_CMP.csv")

#make an MA plot
plotMA(resLFC, alpha = 0.05, main = "MA plot: HSC VS CMP")

#do clustering analysis 

#perform variance stabilizing transformation 
vsd <- vst(dds, blind=FALSE)

#get sample-to-sample distances in a distance matrix
sampleDists <- dist(t(assay(vsd)))

#construct a heatmap from this distance matrix
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(vsd$condition, vsd$type, sep="-")
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)

```

# Pairwise Comparison 2 & 4 - Common Myeloid Progenitor VS Erythroblast

```{r DE-2v4}

#read in libraries
library("DESeq2")
library("apeglm")
library("pheatmap")
library("RColorBrewer")

#read in files associated with all cell lines
cmp.1 <- read.delim("ENCFF623OLU.tsv", header = TRUE, sep = "\t")
cmp.2 <- read.delim("ENCFF691MHW.tsv", header = TRUE, sep = "\t")
ery.1 <- read.delim("ENCFF342WUL.tsv", header = TRUE, sep = "\t")
ery.2 <- read.delim("ENCFF858JHF.tsv", header = TRUE, sep = "\t")

#extract gene id names and TPM columns from each table
count.matrix <- cbind(cmp.1$expected_count, cmp.2$expected_count, ery.1$expected_count, ery.2$expected_count)
#add gene ids as row names to the matrix
rownames(count.matrix) <- cmp.1$gene_id

#make a data frame of column names
col.names <- data.frame(row.names = c("CMP1", "CMP2", "ERY1", "ERY2"), condition = c("CMP", "CMP", "ERY", "ERY"))

#read in the count matrix as a DESeqDataSet
dds <- DESeqDataSetFromMatrix(countData = round(count.matrix),
                              colData = col.names,
                              design = ~ condition) 

#perform DE analysis 
dds <- DESeq(dds)

#find the name of the results you want to perform an LFC shrink on
resultsNames(dds) 

#perform LFC shrinkage
resLFC <- lfcShrink(dds, coef="condition_ERY_vs_CMP", type="apeglm") #documentation indicates apeglm outperforms the normal shrinkage estimator

#subset by only the significant p-values 
resSig <- subset(resLFC, padj < 0.05)

#write to CSV file
write.csv(as.data.frame(resSig), 
          file="DE_CMP_vs_ERY.csv")

#make an MA plot
plotMA(resLFC, alpha = 0.05, main = "MA plot: CMP VS ERY")

#do clustering analysis 

#perform variance stabilizing transformation 
vsd <- vst(dds, blind=FALSE)

#get sample-to-sample distances in a distance matrix
sampleDists <- dist(t(assay(vsd)))

#construct a heatmap from this distance matrix
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(vsd$condition, vsd$type, sep="-")
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)

```

# Pairwise Comparison 3 & 4 - Colony Forming Unit Erythroid VS Erythroblast

```{r DE-3v4}

#read in libraries
library("DESeq2")
library("apeglm")
library("pheatmap")
library("RColorBrewer")

#read in files associated with all cell lines
cfu.1 <- read.delim("ENCFF667IDY.tsv", header = TRUE, sep = "\t")
cfu.2 <- read.delim("ENCFF655LMK.tsv", header = TRUE, sep = "\t")
ery.1 <- read.delim("ENCFF342WUL.tsv", header = TRUE, sep = "\t")
ery.2 <- read.delim("ENCFF858JHF.tsv", header = TRUE, sep = "\t")

#extract gene id names and TPM columns from each table
count.matrix <- cbind(cfu.1$expected_count, cfu.2$expected_count, ery.1$expected_count, ery.2$expected_count)
#add gene ids as row names to the matrix
rownames(count.matrix) <- cfu.1$gene_id

#make a data frame of column names
col.names <- data.frame(row.names = c("CFU1", "CFU2", "ERY1", "ERY2"), condition = c("CFU", "CFU", "ERY", "ERY"))

#read in the count matrix as a DESeqDataSet
dds <- DESeqDataSetFromMatrix(countData = round(count.matrix),
                              colData = col.names,
                              design = ~ condition) 

#perform DE analysis 
dds <- DESeq(dds)

#find the name of the results you want to perform an LFC shrink on
resultsNames(dds) 

#perform LFC shrinkage
resLFC <- lfcShrink(dds, coef="condition_ERY_vs_CFU", type="apeglm") #documentation indicates apeglm outperforms the normal shrinkage estimator

#subset by only the significant p-values 
resSig <- subset(resLFC, padj < 0.05)

#write to CSV file
write.csv(as.data.frame(resSig), 
          file="DE_CFU_vs_ERY.csv")

#make an MA plot
plotMA(resLFC, alpha = 0.05, main = "MA plot: CFU VS ERY")

#do clustering analysis 

#perform variance stabilizing transformation 
vsd <- vst(dds, blind=FALSE)

#get sample-to-sample distances in a distance matrix
sampleDists <- dist(t(assay(vsd)))

#construct a heatmap from this distance matrix
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(vsd$condition, vsd$type, sep="-")
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)

```

# Pairwise Comparison 2 & 3 - Common Myeloid Progenitor VS Colony Forming Unit Erythroid

```{r DE-2v3}

#read in libraries
library("DESeq2")
library("apeglm")
library("pheatmap")
library("RColorBrewer")

#read in files associated with all cell lines
cmp.1 <- read.delim("ENCFF623OLU.tsv", header = TRUE, sep = "\t")
cmp.2 <- read.delim("ENCFF691MHW.tsv", header = TRUE, sep = "\t")
cfu.1 <- read.delim("ENCFF667IDY.tsv", header = TRUE, sep = "\t")
cfu.2 <- read.delim("ENCFF655LMK.tsv", header = TRUE, sep = "\t")

#extract gene id names and TPM columns from each table
count.matrix <- cbind(cmp.1$expected_count, cmp.2$expected_count, cfu.1$expected_count, cfu.2$expected_count)
#add gene ids as row names to the matrix
rownames(count.matrix) <- cfu.1$gene_id

#make a data frame of column names
col.names <- data.frame(row.names = c("CMP1", "CMP2", "CFU1", "CFU2"), condition = c("CMP", "CMP", "CFU", "CFU"))

#read in the count matrix as a DESeqDataSet
dds <- DESeqDataSetFromMatrix(countData = round(count.matrix),
                              colData = col.names,
                              design = ~ condition) 

#perform DE analysis 
dds <- DESeq(dds)

#find the name of the results you want to perform an LFC shrink on
resultsNames(dds) 

#perform LFC shrinkage
resLFC <- lfcShrink(dds, coef="condition_CMP_vs_CFU", type="apeglm") #documentation indicates apeglm outperforms the normal shrinkage estimator

#subset by only the significant p-values 
resSig <- subset(resLFC, padj < 0.05)

#write to CSV file
write.csv(as.data.frame(resSig), 
          file="DE_CMP_vs_CFU.csv")

#make an MA plot
plotMA(resLFC, alpha = 0.05, main = "MA plot: CMP VS CFU")

#do clustering analysis 

#perform variance stabilizing transformation 
vsd <- vst(dds, blind=FALSE)

#get sample-to-sample distances in a distance matrix
sampleDists <- dist(t(assay(vsd)))

#construct a heatmap from this distance matrix
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(vsd$condition, vsd$type, sep="-")
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)

```


